Read ICGEM-format

ioshchepkov opened the pull request:
This is an ICGEM-format parser both for main (2011 year) version and for version 2.0.

The function is somewhat large and more universal than the other input functions. It alone can be used for reading both coefficients as well as their errors. It also computes periodical parts of the coefficients on a given epoch in YYYYMMDD.DD format.

## Checklist

- [x]  remove usecols and add errors attribute, which can be either 'formal', 'calibrated' or None (default)
- [x]  return the coefficients and errors in one call
- [x]  fix docs and typos
- [x] move `other` to `utils`

MarkWieczorek commented:
My suggestion is to keep the old mixed-case naming system only for python-wrapped fortran routines. All python-wrapped fortran routines are found in the `shtools` subpackage, and they are also duplicated in subpackages such as `shio`. 

Once there is a python routine that replaces the functionality of the fortran one, I suggest that we remove it from the relevant subpackage (e.g., `shio`) but keep the old fortran routine in the `shtools` subpackage.

I think that the first priority is to rewrite all the IO routines in `shio`. I dedicated routine for ICGEM is probably needed, but the others (shread, shreadh, shreaderror, shreaderrorh) could probably be combined into a single, relatively generic routine. Based on what we did with the `SHCoeffs.from_file()` method, we might want to make a generic routine like`shread()`, or `read_coeffs()`, etc. and then have an optional arguement that specicies for the format:
```
format : str, optional, default = 'shtools'
    'shtools' format or binary numpy 'npy' format.
```

ioshchepkov commented:
I renamed the main function to `read_icgem_gfc` and moved it to the `pyshtools.shio` subpackage in `icgem.py` and `common.py` submodules. 

Also I created subpackage `utils` for common auxiliary routines. For example, now it has `datetime.py` submodule to store all date and time related routines. Hope to see your opinion on this, because I'm not sure.

MarkWieczorek commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,148 @@
+        elif 'gravity_constant' in header:
+            gravity_constant = float(header['gravity_constant'])
+        else:
+            raise ValueError('No standart gravitational constant in the\
```
typo - No standard

MarkWieczorek commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,148 @@
+        coefficients instead of the coefficients themself. Can be (3, 4) or
+        (5, 6) or (7, 8). Default is (3, 4), i.e. read coefficients.
+    lmax : int, optional
+        Maximum degree to read from file, if None then maximum degree of the
```
Maybe this is clearer? 

Maximum degree to read from the file. If lmax is None, less than 0, or greater than lmax_model, the maximum degree of the model will be used.

MarkWieczorek commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,148 @@
+    filename : str
+        The ascii-formatted filename containing the spherical harmonic coefficients.
+    usecols : sequence, optional
+        Which two columns to read. Can be used to read errors of the
```
It might be a good idea to note what the columns correspond to here.

MarkWieczorek commented on pyshtools/shio/common.py:
```diff
@@ -0,0 +1,28 @@
+"""
```
I wonder if this function should be in utils, or if the single function in utils should be simply in shio. Not sure what the best approach is.

MarkWieczorek commented on pyshtools/shio/common.py:
```diff
@@ -0,0 +1,28 @@
+    """Return sum of the time-variable part of the coefficients
+
+    The formula is:
+    G(t)=G(t0) + trnd*(t-t0 ) +
```
G(t) = G(t0) + trnd*(t-t0)

MarkWieczorek commented on pyshtools/shio/common.py:
```diff
@@ -0,0 +1,28 @@
+
+import numpy as _np
+
+def _derivative(epoch, ref_epoch, trnd, periodic):
```
Does this routine calculate a "derivative"? Not sure if I understand the naming...

MarkWieczorek commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,188 @@
+    """Return sum of the time-variable part of the coefficients
+
+    The formula is:
+    G(t) = G(t0) + trnd*(t-t0 ) +
```
extra space after t0...

ioshchepkov replied:
Fixed.

ioshchepkov commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,188 @@
+    ----------
+    filename : str
+        The ascii-formatted filename containing the spherical harmonic coefficients.
+    errors : str, optional
```
I think the equal sign is misleading here, because what exactly is "default" here from a pythonic point of view? Attribute? Variable?

ioshchepkov commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,188 @@
+def read_icgem_gfc(filename, errors=None, lmax=None, epoch=None):
+    """Read spherical harmonic coefficients from an ICGEM GFC ascii-formatted file.
+
+    This function only reads files with the gravity field spherical
```
It does, the ICGEM-format can also be used for ocean/atmospheric tides, but this function will not be able to read it.

ioshchepkov commented on pyshtools/shio/icgem.py:
```diff
@@ -0,0 +1,188 @@
+            valid_err = ('calibrated', 'formal', 'calibrated_and_formal')
+            if header['errors'] == 'no':
+                raise ValueError('This model has no errors.')
+            elif errors not in valid_err[:-1]:
```
No, 'calibrated_and_formal' is not implemented.

'calibrated', 'formal' and 'calibrated_and_formal' are parameters in the ICGEM formatted file for keyword 'errors' which has nothing to do with the parameter `errors` of the function.

'calibrated' or 'formal' mean there are only two columns with the errors - 5 and 6, which will be read if  `errors="calibrated"` or `errors="formal"` respectively. 

'calibrated_and_formal', in the other hand, means there are four columns in the file: 5, 6, 7 and 8. 5 and 6 stand for calibrated errors and 7 and 8 stand for formal. So, the function will return calibrated errors if `errors="calibrated"` or formal errors if `errors="formal"` as before.

I don't know if returning of both calibrated and formal errors is really needed. If it does, the I will write it.


This part is checking that `errors` has valid values and the given value to this parameter is actually exists in the file.
